bundle/jsky.app.ot/src/main/java/jsky/app/ot/gemini/editor/targetComponent/TelescopePosTableWidget.java

```diff
@@ -350,9 +311,172 @@ private static String extractName(final GuideGroup group) {
+                final Option<SPTarget> primary = gpt.getPrimary();
+
+                // Add all the targets.
+                // The index here is used to generate target names, so begin at 1.
+                gpt.getTargets().zipWithIndex().foreach(tup -> {
+                    final SPTarget target = tup._1();
+                    final int index       = tup._2() + 1;
+
+                    final Option<AgsGuideQuality> quality = guideQuality(ags, guideProbe, target);
+                    final boolean enabled = isPrimaryGroup && primary.exists(target::equals);
+
+                    final Row row = new GuideTargetRow(isActive, quality, enabled, editable, movable, guideProbe,
+                            index, target, baseCoords, when);
+                    rowList.add(row);
+                });
+            });
+
+            return new GroupRow(isPrimaryGroup, false, groupIdx, group, rowList);
+        }
+
+        // Recalculate the quality of all guide stars. If one of them has changed, return a new set of rows.
+        // We return a list of row indices of GroupRows indicating which groups have changed.
+        private Set<Integer> recalculateManualGSQuality(final Option<ObsContext> ctxOpt) {
+            final List<Row>    newRows          = new ArrayList<>();
+            final Set<Integer> groupRowsChanged = new HashSet<>();
+
+            IntStream.range(0, rows.size()).mapToObj(i -> new Pair<>(i, rows.get(i))).forEach(tup -> {
+                final int rowIdx = tup._1();
+                final Row row    = tup._2();
+
+                if (row instanceof GroupRow) {
+                    final IndexedGuideGroup igg = row.group().getValue();
+                    final int gIdx              = igg.index();
+                    final GuideGroup gg         = igg.group();
+
+                    // Recreate the entire group row, and then scan the children to see if any of the qualities are different.
+                    final GroupRow gRowNew = createGroupRow(ctxOpt, gIdx, gg);
+
+                    // Zipping Java lists is too obnoxious.
+                    final ImList<Row> childRowsOld       = DefaultImList.create(row.children());
+                    final ImList<Row> childRowsNew       = DefaultImList.create(gRowNew.children());
+                    final boolean groupChanged           = childRowsOld.zip(childRowsNew).exists(tup2 -> {
+                        final GuideTargetRow gtrOld = (GuideTargetRow) tup2._1();
+                        final GuideTargetRow gtrNew = (GuideTargetRow) tup2._2();
+                        return !gtrOld.quality.exists(q -> gtrNew.quality.exists(q::equals));
+                    });
+
+                    if (groupChanged) {
+                        newRows.add(gRowNew);
+                        groupRowsChanged.add(rowIdx);
+                    } else {
+                        newRows.add(row);
+                    }
+                } else {
+                    newRows.add(row);
```

Comment:
I think maybe I'm a bit confused as to what you mean here.
In one case, `gRowNew` is added, and not `row`.